### ISSUE TITLE
Smart indentation on separated elements should reuse the prior element's starting base indentation

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#53548]
Consider what I'd call some "ideal" smart indentation for the following code samples:

```ts
type Foo =
    | "hello"
    | "world";

{| "indentation": 0 |}
```

```ts
let x = 10,
    y = 2;
{| "indentation": 0 |}
```

```ts
let x = 10,
    y = 2,
    {| "indentation": 4 |}
```

```ts
let obj = {
    x: "hello" +
        "world",
    {| "indentation": 4 |}
}
```

```ts
class C {
    x = "hello" +
        "world";
    {| "indentation": 4 |}
}
```

I've specified the "expected" indentation in each of these cases; however, TypeScript today keeps indentation are the same level as the prior line which is wrong. This is especially annoying for the multi-line union case.

There is some ambiguity when one uses an ASI-less style though; in these cases, I would personally err on the side of expecting the start indentation.

type Foo =
    | "hello"
    | "world"
{| "indentation": 0 |}
```

```ts
let x = 10,
    y = 2
{| "indentation": 0 |}
```

```ts
class C {
    x = "hello" +
        "world"
    {| "indentation": 4 |}
}
```

Then, even for cases where ASI doesn't kick in, I would say we should pick the start element's base, since we'll start doing #52899 eventually

```ts
let obj = {
    x: "hello" +
        "world"
    {| "indentation": 4 |}
}
```

**Comment by DanielRosenwasser** [comment]
@MariaSolOs had a good suggestion which is that (maybe when there's ambiguity over the previous statement is sufficiently "terminated") we should look for the prior line and check its indentation. If it's "less" than the start indentation of the prior statement (if it exists) then we should reset the indentation to that of the prior statement.

**Comment by MariaSolOs** [comment]
I debugged this a bit and didn't see any code in the [smart indenter](https://github.com/microsoft/TypeScript/blob/b6fed21a872feb898971abba009bf4171d9ec871/src/services/formatting/smartIndenter.ts) being used. Is this not the indentation implementation controlling this @DanielRosenwasser?. 

@mjbvz do you know if this indentation behavior is controlled by the "on enter" rules of `language-configuration.json`?

**Comment by DanielRosenwasser** [comment]
You're right. I didn't realize it, but TSServer doesn't even provide smart indentation. I guess this change is going to be editor specific.

**Comment by fatcerberus** [comment]
On a more serious note, this is definitely something I've noticed in day-to-day.  Particularly fun is this case:

```ts
if (a
    && b
    && c)
    {
        // cursor is now here
    } // <-- automatically added
```
So now I get to go up a line, press Home twice, Shift+Down three times, Shift+Tab, then finagle the cursor back inside the braces so I can enter my code.  It's not a showstopper, but is distracting enough to make me lose my train of thought sometimes.

Full-fat VS tends to handle this much better than VSCode, in general.

**Comment by MariaSolOs** [comment]
@fatcerberus I would suggest filing an issue in https://github.com/microsoft/vscode for that :)

**Comment by fatcerberus** [comment]
@MariaSolOs I was under the impression that was handled by the same machinery that *this* issue is about?

**Comment by MariaSolOs** [comment]
@fatcerberus didn't you say that you only experience that in VS Code?
